src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs

```diff
@@ -34,48 +35,65 @@ public HttpApiHandler(DurableTaskExtension config, TraceWriter traceWriter)
             string instanceId,
             OrchestrationClientAttribute attribute)
         {
-            if (this.config.NotificationUrl == null)
-            {
-                throw new InvalidOperationException("Webhooks are not configured");
-            }
-
-            Uri notificationUri = this.config.NotificationUrl;
-
-            // e.g. http://{host}/admin/extensions/DurableTaskExtension?code={systemKey}
-            string hostUrl = request.RequestUri.GetLeftPart(UriPartial.Authority);
-            string baseUrl = hostUrl + notificationUri.AbsolutePath.TrimEnd('/');
-            string instancePrefix = baseUrl + InstancesControllerSegment + WebUtility.UrlEncode(instanceId);
+            this.GetClientResponseLinks(request, instanceId, attribute, out var statusQueryGetUri, out var sendEventPostUri, out var terminatePostUri);
+            return this.CreateResponseMessage(request, instanceId, statusQueryGetUri, sendEventPostUri, terminatePostUri);
+        }
 
-            string taskHub = WebUtility.UrlEncode(attribute.TaskHub ?? config.HubName);
-            string connection = WebUtility.UrlEncode(attribute.ConnectionName ?? config.AzureStorageConnectionStringName ?? ConnectionStringNames.Storage);
 
-            string querySuffix = $"{TaskHubParameter}={taskHub}&{ConnectionParameter}={connection}";
-            if (!string.IsNullOrEmpty(notificationUri.Query))
+        internal async Task<HttpResponseMessage> CreateCheckStatusResponse(
+            HttpRequestMessage request,
+            string instanceId,
+            OrchestrationClientAttribute attribute,
+            TimeSpan timeout,
+            TimeSpan retryInterval)
+        {
+            this.GetClientResponseLinks(request, instanceId, attribute, out var statusQueryGetUri, out var sendEventPostUri, out var terminatePostUri);
+            if (TimeSpan.Compare(timeout, retryInterval) != 1)
             {
-                // This is expected to include the auto-generated system key for this extension.
-                querySuffix += "&" + notificationUri.Query.TrimStart('?');
+                throw new ArgumentException($"Total timeout {timeout.TotalSeconds} should be bigger than retry timeout {retryInterval.TotalSeconds}");
             }
+            this.GetRetryValues(timeout, retryInterval, out var iterationCount, out var leftOverInterval);
```

Comment:
I'm a little concerned about the complexity of this timeout implementation because of how it involves multiple methods, iteration counts, and division.  Could we refactor the timeout logic to be simpler?  Here is a suggestion:

```csharp
Stopwatch stopwatch = Stopwatch.StartNew();
while (true)
{
    var status = await client.GetStatusAsync(instanceId);

    // ...do checks on the status accordingly...

    TimeSpan elapsed = stopwatch.Elapsed;
    if (elapsed < timeout)
    {
        TimeSpan remainingTime = timeout.Subtract(elapsed);
        await Task.Delay(remainingTime > retryInterval ? retryInterval : remainingTime);
    }
    else
    {
        break;
    }
}
```

Reply:
Great suggestion! I will improve it.